lldb/source/Plugins/Protocol/MCP/Resource.cpp

```diff
@@ -0,0 +1,166 @@
+static llvm::Error createStringError(const char *format, Args &&...args) {
+  return llvm::createStringError(
+      llvm::formatv(format, std::forward<Args>(args)...).str());
+}
+
+static llvm::Error createUnsupportedURIError(llvm::StringRef uri) {
+  return llvm::make_error<UnsupportedURI>(uri.str());
+}
+
+protocol::Resource
+DebuggerResourceProvider::GetDebuggerResource(lldb::user_id_t debugger_id) {
+  protocol::Resource resource;
+  resource.uri = llvm::formatv("lldb://debugger/{0}", debugger_id);
+  resource.name = llvm::formatv("debugger {0}", debugger_id);
+  resource.description =
+      llvm::formatv("Information about debugger instance {0}", debugger_id);
+  resource.mimeType = "application/json";
+  return resource;
+}
+
+protocol::Resource
+DebuggerResourceProvider::GetTargetResource(lldb::user_id_t debugger_id,
+                                            lldb::user_id_t target_id) {
+  protocol::Resource resource;
+  resource.uri =
+      llvm::formatv("lldb://debugger/{0}/target/{1}", debugger_id, target_id);
+  resource.name = llvm::formatv("target {0}", target_id);
+  resource.description =
+      llvm::formatv("Information about target {0} in debugger instance {1}",
+                    target_id, debugger_id);
+  resource.mimeType = "application/json";
+  return resource;
+}
+
+std::vector<protocol::Resource> DebuggerResourceProvider::GetResources() const {
+  std::vector<protocol::Resource> resources;
+
+  const size_t num_debuggers = Debugger::GetNumDebuggers();
+  for (size_t i = 0; i < num_debuggers; ++i) {
+    lldb::DebuggerSP debugger_sp = Debugger::GetDebuggerAtIndex(i);
+    if (!debugger_sp)
+      continue;
+    resources.emplace_back(GetDebuggerResource(i));
+
+    TargetList &target_list = debugger_sp->GetTargetList();
+    const size_t num_targets = target_list.GetNumTargets();
+    for (size_t j = 0; j < num_targets; ++j) {
+      lldb::TargetSP target_sp = target_list.GetTargetAtIndex(j);
+      if (!target_sp)
+        continue;
+      resources.emplace_back(GetTargetResource(i, j));
+    }
+  }
+
+  return resources;
+}
+
+llvm::Expected<protocol::ResourceResult>
+DebuggerResourceProvider::ReadResource(llvm::StringRef uri) const {
+  auto [protocol, path] = uri.split("://");
+
+  if (protocol != "lldb")
+    return createUnsupportedURIError(uri);
+
+  llvm::SmallVector<llvm::StringRef, 4> components;
+  path.split(components, '/');
+
+  if (components.size() < 2)
+    return createUnsupportedURIError(uri);
+
+  if (components[0] != "debugger")
+    return createUnsupportedURIError(uri);
+
+  lldb::user_id_t debugger_id;
+  if (components[1].getAsInteger(0, debugger_id))
+    return createStringError("invalid debugger id '{0}': {1}", components[1],
+                             path);
+
+  if (components.size() > 3) {
+    if (components[2] != "target")
+      return createUnsupportedURIError(uri);
+
+    lldb::user_id_t target_id;
+    if (components[3].getAsInteger(0, target_id))
```

Comment:
I wonder if it would be helpful to have support for `lldb://debugger/0/target/selected` as a shorthand for the selected target.

Or in theory `lldb://debugger/0/target` could return info for the selected target.